src/hidpp20.c

```diff
@@ -2412,92 +2421,67 @@ hidpp20_onboard_profiles_parse(struct hidpp20_device *device,
+	if (crc_valid) {
 		for (i = 0; i < profiles->num_profiles; i++) {
 			uint8_t *d = data + 4 * i;
 
-			if (d[0] == 0xFF && d[1] == 0xFF)
+			if (d[HIDPP20_PROFILE_DIR_END1] == HIDPP20_PROFILE_DIR_END_VALUE &&
+			    d[HIDPP20_PROFILE_DIR_END2] == HIDPP20_PROFILE_DIR_END_VALUE)
 				break;
 
 			profiles->profiles[i].address = hidpp_get_unaligned_be_u16(d);
-			if (profiles->profiles[i].address != (dict_address | (i + 1)))
+
+			/* profile address sanity check */
+			if (profiles->profiles[i].address != (HIDPP20_USER_PROFILES_G402 | (i + 1)))
 				hidpp_log_info(&device->base,
-						"profile %d: error in the address: 0x%04x instead of 0x%04x\n",
-						i + 1,
-						profiles->profiles[i].address,
-						dict_address | (i + 1));
-			profiles->profiles[i].enabled = !!d[2];
+					       "profile %d: error in the address: 0x%04x instead of 0x%04x\n",
+					       i + 1,
+					       profiles->profiles[i].address,
+					       HIDPP20_USER_PROFILES_G402 | (i + 1));
+
+			profiles->profiles[i].enabled = !!d[HIDPP20_PROFILE_DIR_ENABLED];
 		}
+	} else {
+		hidpp_log_debug(device, "Profile directory has an invalid CRC... Reading ROM profiles.\n");
+
+		read_userdata = false;
 	}
 
-	/*
-	 * Even if the crc is wrong, we still fetch all the profiles, so as
-	 * to have the current stored values if some profiles are legitimate.
-	 */
-	 for (i = 0; i < profiles->num_profiles; i++) {
-		if (profiles->profiles[i].address == 0x0000) {
-			switch (dict_address) {
-			case 0x0000:
-				profiles->profiles[i].address = i + 1;
-				break;
-			case 0x0100:
-				if (i >= profiles->num_rom_profiles) {
-					profiles->profiles[i].address = dict_address | profiles->num_rom_profiles;
-				} else {
-					profiles->profiles[i].address = dict_address | (i + 1);
-				}
-				break;
-			default:
-				hidpp_log_error(&device->base,
-						"this should never happen: dict address is 0x%04x\n",
-						dict_address);
-				return -EINVAL;
-			}
+	for (i = 0; i < profiles->num_profiles; i++) {
+		if (read_userdata) {
+			hidpp_log_debug(device, "Parsing profile %u\n", i);
+			rc = hidpp20_onboard_profiles_parse_profile(device,
+								    profiles,
+								    i,
+								    true);
+
+			/* on fail to read the user profile fallback to the default profile */
+			if (rc)
+				hidpp_log_debug("Profile %u is bad. Falling back to the ROM settings.\n", i);
+			else
+				continue;
 		}
+
+		/* the number of rom profiles can be different than the number of user profiles
+		   so we if there are not enough rom profiles to populate all the user profiles
+		   we just use the first rom profile */
+		if (i + 1 > profiles->num_rom_profiles)
+			profiles->profiles[i].address = HIDPP20_ROM_PROFILES_G402 + 1;
+		else
+			profiles->profiles[i].address = HIDPP20_ROM_PROFILES_G402 + i + 1;
+
 		rc = hidpp20_onboard_profiles_parse_profile(device,
-							    profiles,
-							    i,
-							    check_crc);
+							profiles,
+							i,
+							false);
+
 		if (rc == -EAGAIN)
 			crc_valid = false;
```

Comment:
What's the point of setting the CRC to invalid here when we know the ROM profile doesn't have a valid CRC, meaning the `rc` error we have here is likely unrelated?